packages-exp/analytics-exp/src/factory.ts

```diff
@@ -0,0 +1,248 @@
+ * all or some of these, depending on the call's `send_to` param and the status
+ * of the dynamic config fetches (see below).
+ */
+export let initializationPromisesMap: {
+  [appId: string]: Promise<string>; // Promise contains measurement ID string.
+} = {};
+
+/**
+ * List of dynamic config fetch promises. In certain cases, wrapped gtag calls
+ * wait on all these to be complete in order to determine if it can selectively
+ * wait for only certain initialization (FID) promises or if it must wait for all.
+ */
+let dynamicConfigPromisesList: Array<Promise<
+  DynamicConfig | MinimalDynamicConfig
+>> = [];
+
+/**
+ * Maps fetched measurementIds to appId. Populated when the app's dynamic config
+ * fetch completes. If already populated, gtag config calls can use this to
+ * selectively wait for only this app's initialization promise (FID) instead of all
+ * initialization promises.
+ */
+const measurementIdToAppId: { [measurementId: string]: string } = {};
+
+/**
+ * Name for window global data layer array used by GA: defaults to 'dataLayer'.
+ */
+let dataLayerName: string = 'dataLayer';
+
+/**
+ * Name for window global gtag function used by GA: defaults to 'gtag'.
+ */
+let gtagName: string = 'gtag';
+
+/**
+ * Reproduction of standard gtag function or reference to existing
+ * gtag function on window object.
+ */
+let gtagCoreFunction: Gtag;
+
+/**
+ * Wrapper around gtag function that ensures FID is sent with all
+ * relevant event and config calls.
+ */
+export let wrappedGtagFunction: Gtag;
+
+/**
+ * Flag to ensure page initialization steps (creation or wrapping of
+ * dataLayer and gtag script) are only run once per page load.
+ */
+let globalInitDone: boolean = false;
+
+/**
+ * For testing
+ * @internal
+ */
+export function resetGlobalVars(
+  newGlobalInitDone = false,
+  newInitializationPromisesMap = {},
+  newDynamicPromises = []
+): void {
+  globalInitDone = newGlobalInitDone;
+  initializationPromisesMap = newInitializationPromisesMap;
+  dynamicConfigPromisesList = newDynamicPromises;
+  dataLayerName = 'dataLayer';
+  gtagName = 'gtag';
+}
+
+/**
+ * For testing
+ * @internal
+ */
+export function getGlobalVars(): {
+  initializationPromisesMap: { [appId: string]: Promise<string> };
+  dynamicConfigPromisesList: Array<
+    Promise<DynamicConfig | MinimalDynamicConfig>
+  >;
+} {
+  return {
+    initializationPromisesMap,
+    dynamicConfigPromisesList
+  };
+}
+
+/**
+ * Configures Firebase Analytics to use custom `gtag` or `dataLayer` names.
+ * Intended to be used if `gtag.js` script has been installed on
+ * this page independently of Firebase Analytics, and is using non-default
+ * names for either the `gtag` function or for `dataLayer`.
+ * Must be called before calling `firebase.analytics()` or it won't
+ * have any effect.
+ *
+ * @public
+ *
+ * @param options - Custom gtag and dataLayer names.
+ */
+export function settings(options: SettingsOptions): void {
```

Comment:
I think I neglected to include it in the API proposal. I changed it to `analyticsSettings`. Should I amend the API proposal?

Reply:
I actually prefer `settings` to `analyticsSettings` since everything is already namespaced by the package name by default. I just thought a name with "verb + thing" pattern like `updateSettings()` feels more natural for a free function.

Yes, we should amend the API proposal. Let's start with `settings()`. We should be able to do everything via emails.